File: src/main/java/io/github/oleksivio/telegram/bot/core/configuration/BotApplicationContext.java
package io.github.oleksivio.telegram.bot.core.configuration;

import io.github.oleksivio.telegram.bot.api.model.annotation.validator.FilterValidator;
import io.github.oleksivio.telegram.bot.core.controller.annotations.holder.CustomValidatorHolder;
import org.springframework.beans.BeansException;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.context.annotation.Configuration;
import io.github.oleksivio.telegram.bot.core.controller.annotations.AnnotationProcessor;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Configuration
public class BotApplicationContext implements ApplicationContextAware {

    private final AnnotationProcessor annotationProcessor;
    private final CustomValidatorHolder customValidatorHolder;

    public BotApplicationContext(AnnotationProcessor annotationProcessor,
                                 CustomValidatorHolder customValidatorHolder) {
        this.annotationProcessor = annotationProcessor;
        this.customValidatorHolder = customValidatorHolder;
    }

    @Override
    public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {

        List<SimpleBean> simpleBeans = new ArrayList<>();
        for (String beanName : applicationContext.getBeanDefinitionNames()) {
            Object instance = applicationContext.getBean(beanName);
            simpleBeans.add(new SimpleBean(beanName, instance));
        }

        for (SimpleBean simpleBean : simpleBeans) {
            if (simpleBean.isFilterValidator()) {
                customValidatorHolder.add(simpleBean.getName(), (FilterValidator) simpleBean.getInstance());
            }
        }

        for (SimpleBean simpleBean : simpleBeans) {
            annotationProcessor.add(simpleBean.getInstance());
        }
    }

    private static class SimpleBean {
        private final String name;
        private final Object instance;

        SimpleBean(String name, Object instance) {
            this.name = name;
            this.instance = instance;
        }

        public String getName() {
            return name;
        }

        Object getInstance() {
            return instance;
        }

        private boolean isFilterValidator() {
            return Objects.equals(instance.getClass().getSuperclass(), FilterValidator.class);
        }
    }
}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/configuration/BotApplicationContext.kt
package io.github.oleksivio.telegram.bot.core.configuration

import io.github.oleksivio.telegram.bot.api.model.annotation.validator.FilterValidator
import io.github.oleksivio.telegram.bot.core.controller.annotations.AnnotationProcessor
import io.github.oleksivio.telegram.bot.core.controller.annotations.holder.CustomValidatorHolder
import org.springframework.beans.BeansException
import org.springframework.context.ApplicationContext
import org.springframework.context.ApplicationContextAware
import org.springframework.context.annotation.Configuration
import kotlin.reflect.full.superclasses

/**
 * Created by oleksivio on 13.03.19 at 19:55
 * Project: telegram-bot-api
 */

@Configuration
class BotApplicationContext(private val annotationProcessor: AnnotationProcessor,
                            private val customValidatorHolder: CustomValidatorHolder) : ApplicationContextAware {

    @Throws(BeansException::class)
    override fun setApplicationContext(applicationContext: ApplicationContext) {

        applicationContext.beanDefinitionNames
                .filter { !it.contains("loggingCodecCustomizer") }
                .filter { !it.contains("jacksonCodecCustomizer") }
                .filter { !it.startsWith("org.springframework") }
                .map { SimpleBean(it, applicationContext.getBean(it)) }
                .forEach {
                    if (it.isFilterValidator) {
                        customValidatorHolder.add(it.name, it.instance as FilterValidator<*>)
                    }
                    // TODO check why is not else condition
                    annotationProcessor.add(it.instance)

                }

    }

    private class SimpleBean constructor(val name: String, val instance: Any) {
        val isFilterValidator: Boolean = instance::class.superclasses.any {
            it == FilterValidator::class
        }
    }

}
